File: problems/696.count-binary-substrings/Solution.cpp
#include <cstring>
#include <iostream>

using namespace std;

class Solution
{
public:
    int countBinarySubstrings(string s)
    {
        int cur = 1, pre = 0, res = 0;
        for (int i = 1; i < s.size(); ++i)
        {
            if (s[i] == s[i - 1])
                ++cur;
            else
            {
                res += min(cur, pre);
                pre = cur;
                cur = 1;
            }
        }
        return res + min(cur, pre);
    }
};

int main()
{
    string str = "00110011";
    Solution s;
    int res = s.countBinarySubstrings(str);
    cout << res;
    return 0;
}

File: problems/696.count-binary-substrings/Solution.py
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        # [2, 2, 2, 2] 分段 0, 1 数量
        s = [len(i) for i in s.replace("01", "0 1").replace("10", "1 0").split()]

        # map the similar index of multiple containers
        return sum(min(a, b) for a, b in zip(s, s[1:]))


s = "00110011"
sol = Solution()
print(sol.countBinarySubstrings(s))